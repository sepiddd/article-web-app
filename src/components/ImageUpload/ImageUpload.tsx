import { useEffect, useState } from "react";
import { Button, Image, message, Space } from "antd";
import { IArticleMode } from "../../types";

interface Props {
  setBase64: (data: any) => void;
  resetForm?: boolean;
  base64?: string;
  mode: IArticleMode;
}

const ImageUpload: React.FC<Props> = ({
  setBase64,
  resetForm,
  base64,
  mode,
}: Props) => {
  const [imagePreview, setImagePreview] = useState<any>("");

  const _handleReaderLoaded = (readerEvt: any) => {
    let binaryString = readerEvt.target.result;
    setBase64(binaryString);
  };

  const photoUpload = (e: any) => {
    e.preventDefault();
    const reader = new FileReader();
    const file = e.target.files[0];
    if (file.size > 4194304) {
      message.error("the image size is larger than 4MB.");
      return false;
    }
    reader.readAsDataURL(file);
    if (reader !== undefined && file !== undefined) {
      reader.onload = _handleReaderLoaded;
      reader.onloadend = () => {
        setImagePreview(reader.result);
      };
    }
  };

  useEffect(() => {
    if (base64) {
      setImagePreview(base64);
    }
  }, [base64]);

  const reset = () => {
    setBase64("");
    setImagePreview("");
  };

  useEffect(() => {
    if (resetForm) {
      reset();
    }
  }, [resetForm]);

  return (
    <Space style={{ width: "100%", justifyContent: "space-between" }}>
      {!imagePreview ? (
        <span>No images ulpoaded yet!</span>
      ) : (
        <Image width={200} src={imagePreview} />
      )}

      {mode !== "read" && (
        <div>
          {imagePreview && (
            <Button
              style={{ marginRight: 10 }}
              onClick={reset}
              shape='round'
              danger
              type='primary'>
              Remove Image
            </Button>
          )}

          <Button
            onChange={() => {}}
            shape='round'
            style={{ alignSelf: "flex-end" }}>
            {imagePreview ? "Change Image" : "Uplaod Image"}
            <input
              type='file'
              id='file'
              accept='.jpeg, .png, .jpg'
              onChange={photoUpload}
              src={imagePreview}
              style={{
                opacity: 0,
                position: "absolute",
                left: 0,
                right: 0,
                bottom: 0,
              }}
            />
          </Button>
        </div>
      )}
    </Space>
  );
};

export default ImageUpload;
