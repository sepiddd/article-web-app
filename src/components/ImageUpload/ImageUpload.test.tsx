import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import ImageUpload from "./ImageUpload";

// const img = "";

// describe("ArticleForm component", () => {
//   test("Matches the snapshot", () => {
//     const readMode = create(
//       <ImageUpload setBase64={() => console.log("readMode")} mode='read' />
//     );
//     const addMode = create(
//       <ImageUpload setBase64={() => console.log("addMode")} mode='add' />
//     );
//     const editMode = create(
//       <ImageUpload setBase64={() => console.log("editMode")} mode='edit' />
//     );

//     const editModeWithBase64 = create(
//       <ImageUpload
//         setBase64={() => console.log("editMode")}
//         mode='edit'
//         base64={img}
//       />
//     );

//     const editModeWithResetForm = create(
//       <ImageUpload
//         setBase64={() => console.log("editMode")}
//         mode='edit'
//         base64={img}
//         resetForm
//       />
//     );

//     expect(readMode.toJSON()).toMatchSnapshot();
//     expect(addMode.toJSON()).toMatchSnapshot();
//     expect(editMode.toJSON()).toMatchSnapshot();
//     expect(editModeWithBase64.toJSON()).toMatchSnapshot();
//     expect(editModeWithResetForm.toJSON()).toMatchSnapshot();
//   });
// });

let container: any = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

// mode==="edit"

it("Image Upload edit mode Without image", () => {
  const fakeData = {
    base64: "",
    setBase64: () => {},
  };
  const setBase64 = jest.fn();

  act(() => {
    render(
      <ImageUpload
        setBase64={setBase64}
        mode='edit'
        base64={fakeData.base64}
      />,
      container
    );
  });

  expect(
    container.querySelector("[data-testid='remove-img']")?.textContent
  ).toBe(undefined);
  expect(container.querySelector("[data-testid='no-image']").textContent).toBe(
    "No images ulpoaded yet!"
  );
  expect(setBase64).toHaveBeenCalledTimes(0);
});

it("Image Upload edit mode With image", () => {
  const fakeData = {
    base64: "x.image",
    setBase64: () => {},
  };
  const setBase64 = jest.fn();

  act(() => {
    render(
      <ImageUpload
        setBase64={fakeData.setBase64}
        mode='edit'
        base64={fakeData.base64}
      />,
      container
    );
  });

  expect(
    container.querySelector("[data-testid='remove-img']")?.textContent
  ).toBe("Remove Image");
  expect(container.querySelector("[data-testid='no-image']")).toBe(null);
  expect(
    container.querySelector("[data-testid='upload-img']").textContent
  ).toBe("Change Image");
  expect(setBase64).toHaveBeenCalledTimes(0);
});

// mode==="read"

it("Image Upload read mode a image", () => {
  act(() => {
    render(
      <ImageUpload setBase64={() => {}} mode='read' base64='' />,
      container
    );
  });
  expect(container.querySelector("[data-testid='remove-img']")).toBe(null);
  expect(container.querySelector("[data-testid='no-image']").textContent).toBe(
    "No images ulpoaded yet!"
  );
  expect(container.querySelector("[data-testid='upload-img']")).toBe(null);
});
