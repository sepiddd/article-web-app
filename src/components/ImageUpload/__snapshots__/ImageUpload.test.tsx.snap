// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleForm component Matches the snapshot 1`] = `
<div
  className="ant-space ant-space-horizontal ant-space-align-center"
  style={
    Object {
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <span>
      No images ulpoaded yet!
    </span>
  </div>
</div>
`;

exports[`ArticleForm component Matches the snapshot 2`] = `
<div
  className="ant-space ant-space-horizontal ant-space-align-center"
  style={
    Object {
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <div
    className="ant-space-item"
    style={
      Object {
        "marginRight": 8,
      }
    }
  >
    <span>
      No images ulpoaded yet!
    </span>
  </div>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <div>
      
      <button
        className="ant-btn ant-btn-round"
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
        type="button"
      >
        <span>
          Uplaod Image
        </span>
        <input
          accept=".jpeg, .png, .jpg"
          id="file"
          onChange={[Function]}
          src=""
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
            }
          }
          type="file"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ArticleForm component Matches the snapshot 3`] = `
<div
  className="ant-space ant-space-horizontal ant-space-align-center"
  style={
    Object {
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <div
    className="ant-space-item"
    style={
      Object {
        "marginRight": 8,
      }
    }
  >
    <span>
      No images ulpoaded yet!
    </span>
  </div>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <div>
      
      <button
        className="ant-btn ant-btn-round"
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
        type="button"
      >
        <span>
          Uplaod Image
        </span>
        <input
          accept=".jpeg, .png, .jpg"
          id="file"
          onChange={[Function]}
          src=""
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
            }
          }
          type="file"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ArticleForm component Matches the snapshot 4`] = `
<div
  className="ant-space ant-space-horizontal ant-space-align-center"
  style={
    Object {
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <div
    className="ant-space-item"
    style={
      Object {
        "marginRight": 8,
      }
    }
  >
    <span>
      No images ulpoaded yet!
    </span>
  </div>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <div>
      
      <button
        className="ant-btn ant-btn-round"
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
        type="button"
      >
        <span>
          Uplaod Image
        </span>
        <input
          accept=".jpeg, .png, .jpg"
          id="file"
          onChange={[Function]}
          src=""
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
            }
          }
          type="file"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ArticleForm component Matches the snapshot 5`] = `
<div
  className="ant-space ant-space-horizontal ant-space-align-center"
  style={
    Object {
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <div
    className="ant-space-item"
    style={
      Object {
        "marginRight": 8,
      }
    }
  >
    <span>
      No images ulpoaded yet!
    </span>
  </div>
  <div
    className="ant-space-item"
    style={Object {}}
  >
    <div>
      
      <button
        className="ant-btn ant-btn-round"
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
        type="button"
      >
        <span>
          Uplaod Image
        </span>
        <input
          accept=".jpeg, .png, .jpg"
          id="file"
          onChange={[Function]}
          src=""
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
            }
          }
          type="file"
        />
      </button>
    </div>
  </div>
</div>
`;
